python/pyspark/sql/pandas/serializers.py

```diff
@@ -1201,11 +1201,89 @@ def generate_data_batches(batches):
 
     def dump_stream(self, iterator, stream):
         """
-        Read through an iterator of (iterator of pandas DataFrame), serialize them to Arrow
-        RecordBatches, and write batches to stream.
+        Read through chained return results from a single partition of handleInputRows.
```

Comment:
Any reason we can't do this in `if key is None:` in `transformWithStateUDF` and `transformWithStateWithInitStateUDF`?

This was my suggestion and I believe you can just do retrieve expired timers and timestamps, and call handleExpiredTimer() with these information, and done. I don't think this  complication is necessary - if we can't do this in `key is None` in some reason, I suspect fixing that would be much easier.

Reply:
I'll wait for the next update about whether my suggestion works or not. I think the complexity would be very different, hence I would like to defer the further review after that.

Reply:
We need to make a tradeoff whether adding the complication in either `serializer.py` or `TransformWithStateInPandasPythonRunner` if we put the above timer handling codes in `if key is None`.
If we put the timer handling logic inside `if key is None`, we will need to call `dump_stream()` again here in `finally `code block: https://github.com/apache/spark/blob/master/python/pyspark/worker.py#L1966. Calling dump_stream() twice means we will need to properly handle how JVM receives batches. Currently we are reusing the `read()` function inside `PythonArrowOutput`, and the reader will end the reading when the batch signals the end here: https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonArrowOutput.scala#L118. Since we are now calling `dump_stream()` twice, We will need to overwrite this function in `TransformWithStateInPandasPythonRunner` and continues reading one more time after receiving end. The extra complexity is that we will also need to properly handle the case where some partitions may not have timer iterator and won't start the additional dump stream writer at all and how we are going to handle exceptions if one of the dump_stream failed. Additionally, we need to set the statefulHandlerState to `TIMER_PROCESSED` after all timer rows are processed so we will need to do some code changes inside `worker.py` to set this properly. I feel like it is better to put all TWS related code changes in one place for better readability. So this means we will need to get the `StatefulProcessorHandlerApiClient` object inside `worker.py` to set the state correctly. This means we will need to have similar code complexity of what we have now in `serializer.py` (return one extra StatefulProcessorHandlerApiClient from `transformWithStateWithInitStateUDF`and deserialize it from `out_iter`). We cannot set the `TIMER_PROCESSED` state in `group_ops.py` because the output rows iterator are not fully consumed there. It is fully consumed after `dump_stream` is called inside `worker.py`.

So either way we will need to deal with extra complexity. I personally think putting timer handling code into `serializer.py` is slightly better because this is more similar to how we are dealing with timer on Scala side - we are chaining the timer output rows after the data handling rows into a single iterator. 

Let me know if you have suggestions on which way is better.

Reply:
https://github.com/HeartSaVioR/spark/commit/f8952b213ba7f2cbfbc78ef145552317812e9f9b

This is the implementation of my suggestion based on 0c5ab3f. I've confirmed that `pyspark.sql.tests.pandas.test_pandas_transform_with_state` passed with this change - I haven't added new tests you've added later though.

I think this is lot much simpler - we just add two markers into input iterator which carries over the mode, and the flow does not change at all. No trick on teeing and chaining iterators, minimum changes on the data structure, etc.

How this works? This is just the same with how we use iterator in Spark in Scala codebase; with iterator in Scala, we pull one entry, process it and produce output, and pull another entry. The generator would have each entry for every grouping key, and then the marker for timer, and then the marker for completion. Each entry will call the function which eventually calls the user function, and the user function is expected to return the iterator, but the logic to produce the iterator should be synchronous (no async and no laziness, otherwise I guess it can even fail without my change).

So when the marker for timer has been evaluated, function calls for all grouping keys must have been already done. Same for the marker for completion. This is same with Scala implementation.

As a side effect, updating the phase is corrected in this commit.

Reply:
If you agree with this, please pick the commit in above. You've already gone through some commits and I can't revert partially by myself.

My fork is public, so you can add my repo and fetch and pull the branch, and cherrypick the commit into this PR branch with merge conflict. I'd recommend you to take whole different way - perform "hard reset" to my commit in this PR branch (`git reset --hard f8952b213ba7f2cbfbc78ef145552317812e9f9b`), and add more commits which are used to address other review comments.

Reply:
Thanks for putting out the commit! I cherry-picked your change and this is now looking much cleaner!